### PR TITLE
Adds Parent property to the repo API

### PR DESCRIPTION
**Pull request by guyzmo**
hop! 👍

cf gogits/go-gogs-client#64

### CODE REVIEW

**Comment by lafriks** [comment]
@guyzmo  you need to create PR to https://github.com/go-gitea/go-sdk to add required field

**Comment by guyzmo** [comment]
done

**Comment by guyzmo** [comment]
I just changed the content, to expose `Repository* BaseRepo` as `'parent'` property, instead of an id. because that's more restful.

**Comment by guyzmo** [comment]
cf [github API](https://developer.github.com/v3/repos/#get), the parent is a full object. The only difference is that the parent does not contain its own parent in the project's tree (to avoid blowing up the connection). Also there's a `source` attribute, that contain the upmost parent project of the tree.

I guess that we can start iteratively, and give the `parent` property as it is currently implemented, at the risk of blowing the connection in case the project tree's branch is huge. Then in a second stage, hack the `parent` property so that it does not contain itself another `parent` property, or at least is set to `null` (and thus `fork=true` + `parent=null` means that there's more).

**Comment by lafriks** [comment]
This code will panic when `BaseRepo` is `nil`. Also I'm not quite sure if `BaseRepo` will be have value at all before `GetBaseRepo` is called

**Comment by guyzmo** [comment]
👌, so we need the code to keep calm when `BaseRepo` is `nil`. How?
And also we need to make sure that `BaseRepo` is valued when the API needs it. How?

I don't know the code base enough to tell how to hack that, so any pointers are welcomed ☺

**Comment by lunny** [comment]
@guyzmo first at all, you should use `govendor update code.gitea.io/sdk` to update the dependency.

**Comment by lafriks** [comment]
Now that dependency is set correctly you can see that unit tests is failing where BaseRepo is nil

**Comment by guyzmo** [comment]
So what can be done, how can we make sure it's 👌 when the pointer is `nil` and also that a tentative is made to resolve the parent when not?

**Comment by lafriks** [comment]
you need to move out assignment up something like:
```go
var parent *api.Repository
if repo.BaseRepo != nil {
    parent = repo.BaseRepo.APIFormat(AccessModeNone)
}
```

**Comment by guyzmo** [comment]
and to make sure the `repo.BaseRepo` contains something?

**Comment by lafriks** [comment]
Then just assign `Parent: parent,`

**Comment by lafriks** [comment]
If you meant to actually load BaseRepo than it should be done outside APIFormat in function from where APIFormat is called

**Comment by guyzmo** [comment]
ok, I've tried something to force load the `BaseRepo`. The build is failing, but it doesn't look like related to that code I wrote, but I might be wrong.

File: models/repo.go [hunk 1]
```diff
@@ -278,6 +278,14 @@ func (repo *Repository) APIFormat(mode AccessMode) *api.Repository {
 		Push:  mode >= AccessModeWrite,
 		Pull:  mode >= AccessModeRead,
 	}
+	parent = nil;
+	err = GetBaseRepo()
+	if err != nil {
+		log.Error(4, "GetRepositoryByID[%d]: %v", id, err)
+	}
+	if repo.BaseRepo != nil {
+		parent := repo.BaseRepo.APIFormat(AccessModeNone)
```
Comment (lafriks) [inline models/repo.go]: it must be `parent = repo.BaseRepo...`

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 1 above)
Comment (guyzmo) [reply models/repo.go]: and I guess `parent := nil` as well. My bad! 👍

**Comment by lafriks** [comment]
But somehow I don't think it will be so easy.
1. This will make loading parents recursively that is not desired
2. Access mode is not correct as you need to give access mode to that repository for current user

Looks like there will be need to load parent and format it outside and set it there

**Comment by guyzmo** [comment]
1. yup, recursive loading isn't great, but as I suggested, it can be Ok, because the edge case where it'll blow the request will be hypra exceptional…
2. err… 🤔?

File: models/repo.go [hunk 2]
```diff
@@ -278,6 +278,14 @@ func (repo *Repository) APIFormat(mode AccessMode) *api.Repository {
 		Push:  mode >= AccessModeWrite,
 		Pull:  mode >= AccessModeRead,
 	}
+	parent := nil;
+	err = GetBaseRepo()
+	if err != nil {
+		log.Error(4, "GetRepositoryByID[%d]: %v", id, err)
+	}
+	if repo.BaseRepo != nil {
```
Comment (bkcsoft) [inline models/repo.go]: Adding `populateParent bool` to the argument-list and then `if populateParent && repo.BaseRepo != nil {` should fix the potential infinite loop :)

**Comment by bkcsoft** [comment]
`BaseRepo` is the repo we forked from right? AFAIK you can't fork a fork so we can't get recursive lookups...

**Comment by guyzmo** [comment]
*you can't fork a fork* → actually you can, but that's 👌

But what you really meant is that a fork cannot be the base of itself (self referencial), or the root and the leaf of a fork tree (cycle). And actually, one project can only be the fork of another project (whereas a project can have several forks). So the graph representation is a tree.

What does that mean? it's that if you don't limit the amount of recursion in the parent property, you might get a super heavy request answer making the HTTP connection timeout, but it cannot create an infinite lookup loop 😉

And to be honest, except for intended foulplay or experimentation, I've yet to see a project tree with more than 3 or 4 levels

File: models/repo.go [hunk 3]
```diff
@@ -272,12 +272,23 @@ func (repo *Repository) APIURL() string {
 
 // APIFormat converts a Repository to api.Repository
 func (repo *Repository) APIFormat(mode AccessMode) *api.Repository {
+	return repo.APIFormatParent(Accessmode, false)
+}
+func (repo *Repository) APIFormatParent(mode AccessMode, isParent) *api.Repository {
```
Comment (bkcsoft) [inline models/repo.go]: `isParent bool`...

**Review by bkcsoft — CHANGES_REQUESTED** [review]
Build breaks to hell and back otherwise 😂

File: models/repo.go (hunk 3 above)
Comment (guyzmo) [reply models/repo.go]: never happened, I don't know what you're talking about 😇 😗

File: models/repo.go [hunk 4]
```diff
@@ -272,12 +272,23 @@ func (repo *Repository) APIURL() string {
...
 		Admin: mode >= AccessModeAdmin,
 		Push:  mode >= AccessModeWrite,
 		Pull:  mode >= AccessModeRead,
 	}
+	parent := nil;
+	err = GetBaseRepo()
+	if err != nil {
+		log.Error(4, "APIFormat issue: %v", id, err)
+	}
+	if repo.BaseRepo != nil {
```
Comment (bkcsoft) [inline models/repo.go]: You're still recursing endlessly 😆

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
LGTM (if build pass :trollface: )

File: models/repo.go [hunk 5]
```diff
@@ -272,12 +272,24 @@ func (repo *Repository) APIURL() string {
...
+func (repo *Repository) innerAPIFormat(mode AccessMode, isParent bool) *api.Repository {
+	var parent *api.Repository
+
 	cloneLink := repo.CloneLink()
 	permission := &api.Permission{
 		Admin: mode >= AccessModeAdmin,
 		Push:  mode >= AccessModeWrite,
 		Pull:  mode >= AccessModeRead,
 	}
+	err := repo.GetBaseRepo()
```
Comment (lafriks) [inline models/repo.go]: if `isParent` is true no need to try to call `GetBaseRepo` as it's result will not be used anyway

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 5 above)
Comment (bkcsoft) [reply models/repo.go]: True :)

File: models/repo.go [hunk 6]
```diff
@@ -272,12 +272,24 @@ func (repo *Repository) APIURL() string {
...
+	if err != nil {
+		log.Error(4, "APIFormat issue: %v", err)
```
Comment (bkcsoft) [inline models/repo.go]: just `"APIFormat: %v", err` ;)

**Review by bkcsoft — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 5 above)
Comment (guyzmo) [reply models/repo.go]: 👌

File: models/repo.go (hunk 6 above)
Comment (guyzmo) [reply models/repo.go]: 👌

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
_now_ it's LGTM :tada:

**Comment by lafriks** [comment]
LGTM

**Review by lafriks — APPROVED** [review]

File: models/repo.go [hunk 7]
```diff
@@ -272,12 +272,26 @@ func (repo *Repository) APIURL() string {
 
 // APIFormat converts a Repository to api.Repository
 func (repo *Repository) APIFormat(mode AccessMode) *api.Repository {
+	return repo.innerAPIFormat(mode, false)
+}
+func (repo *Repository) innerAPIFormat(mode AccessMode, isParent bool) *api.Repository {
```
Comment (lunny) [inline models/repo.go]: blank line

Comment (guyzmo) [reply models/repo.go]: done 👌

**Comment by lunny** [comment]
let L-G-T-M work

**Review by appleboy — APPROVED** [review]

**Pull request merged**